### DISCUSSION TITLE
Overriding $font-family-base with var (using next/font)

### DISCUSSION THREAD

**Discussion by marceloverdijk** [twbs/bootstrap#39244] [Q&A] [answered]
I'm trying to override the `$font-family-base` using next/font.

I did this already for `$headings-font-family` which worked, but for `$font-family-base` I don't get it working.

I used https://nextjs.org/docs/pages/building-your-application/optimizing/fonts to setup the fonts like:

```
import type { AppProps } from 'next/app'
import { Open_Sans, Poppins } from 'next/font/google'
import '@/styles/styles.scss'
import Layout from '@/components/Layout'

const open_sans = Open_Sans({
  display: 'swap' ,
  subsets: ['latin'],
  variable: '--font-open-sans',
})

const poppins = Poppins({
  display: 'swap',
  subsets: ['latin'],
  variable: '--font-poppins',
  weight: ['400', '500', '700'], 
})

export default function App({ Component, pageProps }: AppProps) {
  return (
    <Layout>
      <div className={`${open_sans.variable} ${poppins.variable}`}>
        <Component {...pageProps} />
      </div>
    </Layout>
  )
}
```

It includes both the Open Sans and Poppins fonts which are made available using the defined variables.

Now, I setup Bootstrap like:

```
// https://getbootstrap.com/docs/5.3/customize/sass/

// 1. Include functions first (so you can manipulate colors, SVGs, calc, etc)
@import "bootstrap/scss/functions";

// 2. Include any default variable overrides here

$enable-negative-margins: true;
$font-family-base:        var(--font-open-sans), sans-serif;
$headings-font-family:    var(--font-poppins), sans-serif;

// 3. Include remainder of required Bootstrap stylesheets (including any separate color mode stylesheets)
@import "bootstrap/scss/variables";
@import "bootstrap/scss/variables-dark";

// 4. Include any default map overrides here

// 5. Include remainder of required parts
@import "bootstrap/scss/maps";
@import "bootstrap/scss/mixins";
@import "bootstrap/scss/root";

// 6. Optionally include any other parts as needed
@import "bootstrap/scss/bootstrap";
// TODO: Only import what is really needed

// 7. Optionally include utilities API last to generate classes based on the Sass map in `_utilities.scss`
@import "bootstrap/scss/utilities/api";

// 8. Add additional custom code here

```

So I override the `$font-family-base` and `$headings-font-family` in step 2.

For the `$headings-font-family` this works as Poppins is used,
but for the `$font-family-base`it does not work as I for some reason get Times as font... 

I don't understand what I'm doing wrong here.

**Comment by marceloverdijk** [comment]
Looking at the generated code this looks good:

<img width="789" alt="image" src="https://github.com/twbs/bootstrap/assets/197004/89868078-50b0-47a0-a4f4-342496b3f935">

and the headings use the correct font as well:

<img width="790" alt="image" src="https://github.com/twbs/bootstrap/assets/197004/aa29596d-3256-4c65-9678-3dba6fc5db59">

**Comment by julien-deramond** [comment]
I haven't checked the whole issue that you have, but by reading the code, there's something that could cause an issue: you import the whole `bootstrap.scss` at step 6. 

```scss
// 6. Optionally include any other parts as needed
@import "bootstrap/scss/bootstrap";
```

IDK if it is the problem here, but I can't guarantee that it'll work either because it's like reimporting some files.
Just to be sure, you can try for now to import the files listed in step 6 at https://getbootstrap.com/docs/5.3/customize/sass/.
If it still doesn't work, it means that there's another issue.

**Reply by marceloverdijk** [reply]
Thx for the feedback and helping out @julien-deramond 👍 

I've tried as you suggested using:

```
// https://getbootstrap.com/docs/5.3/customize/sass/

// 1. Include functions first (so you can manipulate colors, SVGs, calc, etc)
@import "bootstrap/scss/functions";

// 2. Include any default variable overrides here

$enable-negative-margins: true;
$font-family-base:        var(--font-open-sans), sans-serif;
$headings-font-family:    var(--font-poppins), sans-serif;

// 3. Include remainder of required Bootstrap stylesheets (including any separate color mode stylesheets)
@import "bootstrap/scss/variables";
@import "bootstrap/scss/variables-dark";

// 4. Include any default map overrides here

// 5. Include remainder of required parts
@import "bootstrap/scss/maps";
@import "bootstrap/scss/mixins";
@import "bootstrap/scss/root";

// 6. Optionally include any other parts as needed
// @import "bootstrap/scss/bootstrap";
@import "bootstrap/scss/utilities";
@import "bootstrap/scss/reboot";
@import "bootstrap/scss/type";
@import "bootstrap/scss/images";
@import "bootstrap/scss/containers";
@import "bootstrap/scss/grid";
@import "bootstrap/scss/tables";
@import "bootstrap/scss/forms";
@import "bootstrap/scss/buttons";
@import "bootstrap/scss/transitions";
@import "bootstrap/scss/dropdown";
@import "bootstrap/scss/button-group";
@import "bootstrap/scss/nav";
@import "bootstrap/scss/navbar";
@import "bootstrap/scss/card";
@import "bootstrap/scss/accordion";
@import "bootstrap/scss/breadcrumb";
@import "bootstrap/scss/pagination";
@import "bootstrap/scss/badge";
@import "bootstrap/scss/alert";
@import "bootstrap/scss/progress";
@import "bootstrap/scss/list-group";
@import "bootstrap/scss/close";
@import "bootstrap/scss/toasts";
@import "bootstrap/scss/modal";
@import "bootstrap/scss/tooltip";
@import "bootstrap/scss/popover";
@import "bootstrap/scss/carousel";
@import "bootstrap/scss/spinners";
@import "bootstrap/scss/offcanvas";
@import "bootstrap/scss/placeholders";
@import "bootstrap/scss/helpers";

// 7. Optionally include utilities API last to generate classes based on the Sass map in `_utilities.scss`
@import "bootstrap/scss/utilities/api";

// 8. Add additional custom code here
```

in step 6 I have now imported all files that are also imported from `bootstrap/scss/bootstrap` except the ones already imported in previous steps.

But the result is unfortunately the same.

Would it help if I create a simple repo reproducing this issue?


Secondly I was importing `bootstrap/scss/bootstrap` as I noticed many other (commercial) templates - e.g. bought via getbootstrap/themes - do the same; but indeed this is maybe not correct.
On the other hand importing all the separate files as I did now in step 6 might also introduce an issue later if e.g. later a new component is introduced. Would it be an idea to create a separate include file for this in Bootstrap itself, so either Bootstrap itself can import it, but als users when customizing import that in stop 6 as suggested in https://getbootstrap.com/docs/5.3/customize/sass/.

**Reply by julien-deramond** [reply]
> Would it help if I create a simple repo reproducing this issue?

Yep, it would be really helpful if not too complicated for you to create 🙏 
If I tried to create it on my own, I would not have exactly the same way to implement/declare/import things

**Reply by marceloverdijk** [reply]
Sure I will create it this evening. I have a small starter project I can share easily.

**Reply by marceloverdijk** [reply]
Here is the repo: https://github.com/marceloverdijk/bootstrap-39244
it's just nextjs, bootstrap and react-bootstrap basically, and a root page containing a header (Poppins expected and rendered) and a paragraph (Open Sans expected but not rendered).

Note: I followed this: https://nextjs.org/docs/pages/building-your-application/optimizing/fonts

**Reply by julien-deramond** [reply, marked as answer]
Thanks for the repro, it helped a lot to understand the context.

So I tried to do the following to double-check that Sass variable override works well. In your code, I used first `$font-family-base: var(--font-open-sans, Arial), sans-serif;`. And the rendering of the paragraph fallbacks to Arial. So this part is OK.

So now, why does it fallbacks to Arial? Probably because `--font-open-sans` is not defined. It is defined in your code, but (IDK Next but this is what seems to happen) in the scope of the `<div>`, at least it is defined like this:

```css
.__variable_cb35e5 {
	--font-poppins: '__Poppins_cb35e5', '__Poppins_Fallback_cb35e5';
}
```

I'm not 100% sure of the root cause, but the issue is probably coming from the fact that our `--bs-body-font-family` is defined at the `:root` level and `--font-open-sans` is not "accessible" from there. Something like that.

I managed to make it work in your project, but this is maybe not the direction you'd like to take depending on several factors linked to your project. But I hope it can help:

In `_app.tsx` you could load your fonts at the `:root` level too:

```tsx
export default function App({ Component, pageProps }: AppProps) {
  return (
    <>
      <style jsx global>{`
        :root {
          --font-poppins: ${poppins.style.fontFamily};
          --font-open-sans: ${open_sans.style.fontFamily};
        }
      `}</style>
      <Layout>
        <div className={`${open_sans.variable} ${poppins.variable}`}>
          <Component {...pageProps} />
        </div>
      </Layout>
    </>
  );
}
```

**Reply by marceloverdijk** [reply]
Thanks a lot @julien-deramond for your thorough investigation and the explanation; really appreciated. This should help me go forward.

**Reply by marceloverdijk** [reply]
PS: what do you think about the other suggestion:

> Secondly I was importing bootstrap/scss/bootstrap as I noticed many other (commercial) templates - e.g. bought via getbootstrap/themes - do the same; but indeed this is maybe not correct.
On the other hand importing all the separate files as I did now in step 6 might also introduce an issue later if e.g. later a new component is introduced. Would it be an idea to create a separate include file for this in Bootstrap itself, so both Bootstrap itself can import it, but als users when customizing import that in stop 6 as suggested in https://getbootstrap.com/docs/5.3/customize/sass/.

**Reply by julien-deramond** [reply]
> might also introduce an issue later if e.g. later a new component is introduced

If a component has just been introduced to Bootstrap, you wouldn't need to import it since you don't use it yet (except if I'm missing some use cases here, please tell me). As soon as you start using it, you can add it to your Sass compilation. By only selecting what's needed, the size of your CSS will be lighter.

However, the first time you need to create the list of the components to import in your Sass file, it's difficult to find the information which is only in our own code and not in the documentation. Thinking out loud here but maybe we could reflect the content of our `bootstrap.scss` file in the documentation with some explanations; and you would find the complete list of components to import.
